includes/chart.js

```diff
@@ -0,0 +1,330 @@
+            var offsetX = 275; // Adjust this value as needed (!!! This is a little temporary hack !!!)
+            var offsetY = 225; // Adjust this value as needed (!!! This is a little temporary hack !!!)
+            // Calculate the center of the bounding box with the offset
+            sourceX = sourceBBox.left + sourceBBox.width / 2 - offsetX;
+            sourceY = sourceBBox.top + sourceBBox.height / 2 - offsetY;
+        } else if (sourceNode.tagName === 'rect') {
+            sourceX = parseFloat(sourceNode.getAttribute("x")) + parseFloat(sourceNode.getAttribute("width")) / 2;
+            sourceY = parseFloat(sourceNode.getAttribute("y")) + parseFloat(sourceNode.getAttribute("height")) / 2;
+        }
+
+        if (targetNode.tagName === 'ellipse') {
+            targetX = parseFloat(targetNode.getAttribute("cx"));
+            targetY = parseFloat(targetNode.getAttribute("cy"));
+        } else if (targetNode.tagName === 'polygon') {
+            var targetBBox = targetNode.getBoundingClientRect();
+            targetX = targetBBox.left + targetBBox.width / 2;
+            targetY = targetBBox.top + targetBBox.height / 2;
+        } else if (targetNode.tagName === 'rect') {
+            targetX = parseFloat(targetNode.getAttribute("x")) + parseFloat(targetNode.getAttribute("width")) / 2;
+            targetY = parseFloat(targetNode.getAttribute("y")) + parseFloat(targetNode.getAttribute("height")) / 2;
+        }
+
+        var line = document.querySelector(`line[data-source="${rowID}_${link.source}"][data-target="${rowID}_${link.target}"]`);
+        if (!line) {
+            line = document.createElementNS("http://www.w3.org/2000/svg", "line");
+            line.setAttribute("stroke", "gray");
+            line.setAttribute("stroke-width", "2");
+            line.setAttribute("data-source", `${rowID}_${link.source}`);
+            line.setAttribute("data-target", `${rowID}_${link.target}`);
+        }
+
+        line.setAttribute("x1", sourceX);
+        line.setAttribute("y1", sourceY);
+        line.setAttribute("x2", targetX);
+        line.setAttribute("y2", targetY);
+
+        // Find the first shape node (ellipse, polygon, or rect)
+        var firstShape = document.querySelector("svg ellipse, svg polygon, svg rect");
+
+        // Insert the line before the first shape node
+        if (firstShape) {
+            firstShape.parentNode.insertBefore(line, firstShape);
+        } else {
+            // If no shape is found, append the line to the end of the SVG
+            document.querySelector("svg").appendChild(line);
+        }
+    });
+}
+
+
+// Function to calculate polygon centroid
+function calculatePolygonCentroid(polygon) {
+    var points = polygon.points;
+    var numPoints = points.numberOfItems;
+    var x = 0, y = 0;
+
+    for (var i = 0; i < numPoints; i++) {
+        x += points.getItem(i).x;
+        y += points.getItem(i).y;
+    }
+
+    x /= numPoints;
+    y /= numPoints;
```

Comment:
Why is this happening? Descriptive var names and occasionally more verbosity could help readability.

Reply:
x/y names made more explicit